



## Importing libraries and dataset

In [2]:
pip install plotly-express  

     |████████████████████████████████| 26.5 MB 10.1 MB/s 
Note: you may need to restart the kernel to use updated packages.


In [4]:
import pandas as pd
import numpy as np
import plotly.express as px
import math 
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import warnings
warnings.filterwarnings('ignore')
from sqlalchemy import create_engine


df = pd.read_csv('fifa21_male2.csv')

## Exploring data

In [5]:
# This syntax is for displaying all columns and 200 rows from dataset

pd.set_option('display.max_rows', 200)
pd.set_option('display.max_columns', None)

In [6]:
df

ID             Name  Age  OVA            Nationality  \
0           2      G. Pasquale   33   69                  Italy   
1          16      Luis García   37   71                  Spain   
2          27          J. Cole   33   71                England   
3          36         D. Yorke   36   68  Trinidad &amp; Tobago   
4          41          Iniesta   36   81                  Spain   
...       ...              ...  ...  ...                    ...   
17120  259088       A. Medioub   22   64                 France   
17121  259090       W. Rickard   18   56                  Wales   
17122  259091       C. Barrett   18   56                England   
17123  259101   J. Gazibegović   20   62     Bosnia Herzegovina   
17124  259105  Pablo Rodríguez   18   59                  Spain   

                Club  BOV   BP      Position  \
0            Udinese   71  LWB            LM   
1          KAS Eupen   70   CM    CM CAM CDM   
2      Coventry City   71  CAM  CAM RM RW LM   
3         Sunderland   70   ST           NaN   
4        Vissel Kobe   82  CAM        CM CAM   
...              ...  ...  ...           ...   
17120     CD Tondela   66   CB            CB   
17121        Burnley   58  CAM        CAM CM   
17122        Burnley   56   RB            RB   
17123  SK Sturm Graz   62   RB         RB LB   
17124          Lecce   61   ST            ST   

                                            Player Photo  \
0      https://cdn.sofifa.com/players/000/002/16_120.png   
1      https://cdn.sofifa.com/players/000/016/19_120.png   
2      https://cdn.sofifa.com/players/000/027/16_120.png   
3      https://cdn.sofifa.com/players/000/036/09_120.png   
4      https://cdn.sofifa.com/players/000/041/20_120.png   
...                                                  ...   
17120  https://cdn.sofifa.com/players/259/088/20_120.png   
17121  https://cdn.sofifa.com/players/259/090/20_120.png   
17122  https://cdn.sofifa.com/players/259/091/20_120.png   
17123  https://cdn.sofifa.com/players/259/101/20_120.png   
17124  https://cdn.sofifa.com/players/259/105/20_120.png   

                                              Club Logo  \
0          https://cdn.sofifa.com/teams/55/light_60.png   
1        https://cdn.sofifa.com/teams/2013/light_60.png   
2        https://cdn.sofifa.com/teams/1800/light_60.png   
3         https://cdn.sofifa.com/teams/106/light_60.png   
4      https://cdn.sofifa.com/teams/101146/light_60.png   
...                                                 ...   
17120  https://cdn.sofifa.com/teams/112516/light_60.png   
17121    https://cdn.sofifa.com/teams/1796/light_60.png   
17122    https://cdn.sofifa.com/teams/1796/light_60.png   
17123     https://cdn.sofifa.com/teams/209/light_60.png   
17124     https://cdn.sofifa.com/teams/347/light_60.png   

                                    Flag Photo  POT  \
0          https://cdn.sofifa.com/flags/it.png   69   
1          https://cdn.sofifa.com/flags/es.png   71   
2      https://cdn.sofifa.com/flags/gb-eng.png   71   
3          https://cdn.sofifa.com/flags/tt.png   82   
4          https://cdn.sofifa.com/flags/es.png   81   
...                                        ...  ...   
17120      https://cdn.sofifa.com/flags/fr.png   69   
17121  https://cdn.sofifa.com/flags/gb-wls.png   66   
17122  https://cdn.sofifa.com/flags/gb-eng.png   67   
17123      https://cdn.sofifa.com/flags/ba.png   73   
17124      https://cdn.sofifa.com/flags/es.png   74   

                       Team & Contract Height  Weight   foot  Growth  \
0                  Udinese 2008 ~ 2016   6'0"  181lbs   Left       0   
1                KAS Eupen 2014 ~ 2019  5'10"  143lbs  Right       0   
2            Coventry City 2016 ~ 2020   5'9"  161lbs  Right       0   
3                      Sunderland 2009  5'11"  165lbs  Right      14   
4              Vissel Kobe 2018 ~ 2021   5'7"  150lbs  Right       0   
...                                ...    ...     ...    ...     ...   
17120  CD Tondela Jun 30, 2021 On 

In [7]:
df.shape

(17125, 107)

In [8]:
df.describe().T

count           mean           std    min       25%  \
ID                17125.0  219388.716204  37499.197507    2.0  204082.0   
Age               17125.0      25.272934      4.942665   16.0      21.0   
OVA               17125.0      66.965022      6.864329   38.0      62.0   
BOV               17125.0      67.900204      6.637538   42.0      64.0   
POT               17125.0      72.489810      5.769949   47.0      69.0   
Growth            17125.0       5.524788      5.787539   -1.0       0.0   
Attacking         17125.0     258.537752     72.255687   33.0     232.0   
Crossing          17125.0      51.621314     17.875715    6.0      41.0   
Finishing         17125.0      47.956672     19.391547    3.0      33.0   
Heading Accuracy  17125.0      53.598832     16.947164    5.0      46.0   
Short Passing     17125.0      60.507620     13.960635    7.0      56.0   
Volleys           17067.0      45.005742     17.729483    4.0      32.0   
Skill             17125.0     266.607591     76.809344   43.0     232.0   
Dribbling         17125.0      57.846423     18.047250    5.0      53.0   
Curve             17067.0      49.574676     18.188344    4.0      37.0   
FK Accuracy       17125.0      44.387270     17.504128    5.0      32.0   
Long Passing      17125.0      54.324321     14.856911    9.0      45.0   
Ball Control      17125.0      60.642803     15.873647    5.0      57.0   
Movement          17125.0     322.693664     54.933593  113.0     294.0   
Acceleration      17125.0      65.445898     14.657889   12.0      58.0   
Sprint Speed      17125.0      65.443036     14.437377   11.0      59.0   
Agility           17067.0      64.602273     14.372443   14.0      57.0   
Reactions         17125.0      62.924847      9.037279   24.0      57.0   
Balance           17067.0      64.715591     14.031563   17.0      57.0   
Power             17125.0     302.414774     49.380576  128.0     272.0   
Shot Power        17125.0      59.711416     13.417931   12.0      50.0   
Jumping           17067.0      65.170212     11.918063   22.0      58.0   
Stamina           17125.0      63.309606     15.313272   11.0      56.0   
Strength          17125.0      65.308555     12.627212   16.0      58.0   
Long Shots        17125.0      49.135708     19.044862    4.0      35.0   
Mentality         17125.0     261.922803     62.913935   50.0     235.0   
Aggression        17125.0      56.997839     17.021619    9.0      45.0   
Interceptions     17118.0      47.085582     20.846705    4.0      26.0   
Positioning       17118.0      52.372766     19.094432    2.0      43.0   
Vision            17067.0      55.443605     13.730080   10.0      47.0   
Penalties         17125.0      50.251445     15.696363    7.0      40.0   
Composure         16702.0      59.940965     11.751111   12.0      53.0   
Defending         17125.0     141.469898     61.595180   17.0      84.0   
Marking           17125.0      47.246949     20.182548    3.0      29.0   
Standing Tackle   17125.0      48.279358     21.483597    5.0      28.0   
Sliding Tackle    17067.0      46.099725     21.185154    6.0      25.0   
Goalkeeping       17125.0      77.609693     81.136603    5.0      48.0   
GK Diving         17125.0      15.600467     16.828637    1.0       8.0   
GK Handling       17125.0      15.484672     16.143679    1.0       8.0   
GK Kicking        17125.0      15.472058     15.997328    1.0       8.0   
GK Positioning    17125.0      15.514277     16.406064    1.0       8.0   
GK Reflexes       17125.0      15.738628     17.130908    1.0       8.0   
Total Stats       17125.0    1631.256175    260.357024  731.0    1492.0   
Base Stats        17125.0     361.372146     40.252290  228.0     333.0   
PAC               17125.0      68.091620     11.147116   25.0      62.0   
SHO               17125.0      54.967299     13.877605   16.0      46.0   
PAS               17125.0      58.926540     10.186728   25.0      52.0   
DRI               17125.0      64.209869     

In [9]:
#Standardizing column headers

df.columns = [column.lower().replace(' ', '_') for column in df.columns]

In [10]:
#Set 'id' column as index(optional)

df.set_index('id',inplace=True)

In [11]:
df.dtypes

name                 object
age                   int64
ova                   int64
nationality          object
club                 object
bov                   int64
bp                   object
position             object
player_photo         object
club_logo            object
flag_photo           object
pot                   int64
team_&_contract      object
height               object
weight               object
foot                 object
growth                int64
joined               object
loan_date_end        object
value                object
wage                 object
release_clause       object
contract             object
attacking             int64
crossing              int64
finishing             int64
heading_accuracy      int64
short_passing         int64
volleys             float64
skill                 int64
dribbling             int64
curve               float64
fk_accuracy           int64
long_passing          int64
ball_control          int64
movement            

### Checking NaN values

In [12]:
# For displaying only columns with NaN values

column_with_nan = df.columns[df.isnull().any()]
for column in column_with_nan:
     print(column, df[column].isnull().sum())


club 23
position 413
club_logo 23
joined 58
loan_date_end 16215
volleys 58
curve 58
agility 58
balance 58
jumping 58
interceptions 7
positioning 7
vision 58
composure 423
sliding_tackle 58
a/w 89
d/w 89


### Splitting data into numerical and categorical

In [13]:
X_cat = df.select_dtypes(include = object)
X_num = df.select_dtypes(['float64','int64'])

In [14]:
for col in X_cat.columns:
    print(f'Number of unique values in {col}: {X_cat[col].nunique()}')  

Number of unique values in name: 16176
Number of unique values in nationality: 167
Number of unique values in club: 917
Number of unique values in bp: 15
Number of unique values in position: 622
Number of unique values in player_photo: 17125
Number of unique values in club_logo: 918
Number of unique values in flag_photo: 167
Number of unique values in team_&_contract: 9822
Number of unique values in height: 21
Number of unique values in weight: 57
Number of unique values in foot: 2
Number of unique values in joined: 1953
Number of unique values in loan_date_end: 38
Number of unique values in value: 216
Number of unique values in wage: 142
Number of unique values in release_clause: 1200
Number of unique values in contract: 434
Number of unique values in w/f: 5
Number of unique values in sm: 5
Number of unique values in a/w: 3
Number of unique values in d/w: 3
Number of unique values in ir: 5
Number of unique values in hits: 593
Number of unique values in ls: 278
Number of unique values 

# ## Cleaning Null values from some numerical data(replacing with mean)

In [15]:
##These columns have quite few nulls.I would not prefer drop them,so ifill NaN values with mean of columns

df['volleys'].replace({np.NaN:df['volleys'].mean()},inplace=True)
df['curve'].replace({np.NaN:df['curve'].mean()},inplace=True)
df['agility'].replace({np.NaN:df['agility'].mean()},inplace=True)
df['balance'].replace({np.NaN:df['balance'].mean()},inplace=True)
df['jumping'].replace({np.NaN:df['jumping'].mean()},inplace=True)
df['interceptions'].replace({np.NaN:df['interceptions'].mean()},inplace=True)
df['positioning'].replace({np.NaN:df['positioning'].mean()},inplace=True)
df['composure'].replace({np.NaN:df['composure'].mean()},inplace=True)
df['sliding_tackle'].replace({np.NaN:df['sliding_tackle'].mean()},inplace=True)

### Creating a function for cleaning to apply to columns with currency

In [16]:
def cleaning_currency(x):
    x = str(x).replace('€', '')
    if('M' in str(x)):
        x = str(x).replace('M', '')
        x = float(x) * 1000000
    elif('K' in str(x)):
        x = str(x).replace('K', '')
        x = float(x) * 1000
    return float(x)

### Cleaning  some columns

In [17]:
#Cleaning columns with currency 

df['release_clause'] = round(df['release_clause'].apply(cleaning_currency))

df['wage'] = round(df['wage'].apply(cleaning_currency))

df['value'] = round(df['value'].apply(cleaning_currency))


##Cleaning 'weight' column(As all the values are in 'lbs' we can perform following convertion)

df['weight'] = df['weight'].str.replace('lbs','')

df['weight'] = df['weight'].astype(float)

df['weight'] = round(df['weight']*0.45)   # converting lbs to kg


# Cleaning 'height' column

def convert_height(x):   
    x = str(x).replace('"', '')  ## Removes " sign
    h_ft = int(x.split("'")[0])    # Splits by ' sign and takes first element of list  
    h_inch = int(x.split("'")[1])  # takes second element of list 
    h_inch += h_ft * 12
    h_cm = round(h_inch * 2.54)       
    return int(h_cm)                 # converts value with cm


# Apply the function

df['height'] = df['height'].apply(convert_height)



# Cleaning w/f,sm,ir columns from star signs

df['w/f'].to_string()
df['w/f'] = df['w/f'].str.replace('★','')


df['sm'].to_string()
df['sm'] = df['sm'].str.replace('★', '')



df['ir'].to_string()
df['ir'] = df['ir'].str.replace('★', '')



# Cleaning 'contract' column

split=df['contract'].str.split(r'[~,]',n = 1,expand = True)  ## We split dates separated by ~ or commas
df['contract_start']= split[0] 
df['contract_end']= split[1]
df.drop(columns=['contract'])




name  age  ova            nationality           club  bov  \
id                                                                             
2           G. Pasquale   33   69                  Italy        Udinese   71   
16          Luis García   37   71                  Spain      KAS Eupen   70   
27              J. Cole   33   71                England  Coventry City   71   
36             D. Yorke   36   68  Trinidad &amp; Tobago     Sunderland   70   
41              Iniesta   36   81                  Spain    Vissel Kobe   82   
...                 ...  ...  ...                    ...            ...  ...   
259088       A. Medioub   22   64                 France     CD Tondela   66   
259090       W. Rickard   18   56                  Wales        Burnley   58   
259091       C. Barrett   18   56                England        Burnley   56   
259101   J. Gazibegović   20   62     Bosnia Herzegovina  SK Sturm Graz   62   
259105  Pablo Rodríguez   18   59                  Spain          Lecce   61   

         bp      position                                       player_photo  \
id                                                                             
2       LWB            LM  https://cdn.sofifa.com/players/000/002/16_120.png   
16       CM    CM CAM CDM  https://cdn.sofifa.com/players/000/016/19_120.png   
27      CAM  CAM RM RW LM  https://cdn.sofifa.com/players/000/027/16_120.png   
36       ST           NaN  https://cdn.sofifa.com/players/000/036/09_120.png   
41      CAM        CM CAM  https://cdn.sofifa.com/players/000/041/20_120.png   
...     ...           ...                                                ...   
259088   CB            CB  https://cdn.sofifa.com/players/259/088/20_120.png   
259090  CAM        CAM CM  https://cdn.sofifa.com/players/259/090/20_120.png   
259091   RB            RB  https://cdn.sofifa.com/players/259/091/20_120.png   
259101   RB         RB LB  https://cdn.sofifa.com/players/259/101/20_120.png   
259105   ST            ST  https://cdn.sofifa.com/players/259/105/20_120.png   

                                               club_logo  \
id                                                         
2           https://cdn.sofifa.com/teams/55/light_60.png   
16        https://cdn.sofifa.com/teams/2013/light_60.png   
27        https://cdn.sofifa.com/teams/1800/light_60.png   
36         https://cdn.sofifa.com/teams/106/light_60.png   
41      https://cdn.sofifa.com/teams/101146/light_60.png   
...                                                  ...   
259088  https://cdn.sofifa.com/teams/112516/light_60.png   
259090    https://cdn.sofifa.com/teams/1796/light_60.png   
259091    https://cdn.sofifa.com/teams/1796/light_60.png   
259101     https://cdn.sofifa.com/teams/209/light_60.png   
259105     https://cdn.sofifa.com/teams/347/light_60.png   

                                     flag_photo  pot  \
id                                                     
2           https://cdn.sofifa.com/flags/it.png   69   
16          https://cdn.sofifa.com/flags/es.png   71   
27      https://cdn.sofifa.com/flags/gb-eng.png   71   
36          https://cdn.sofifa.com/flags/tt.png   82   
41          https://cdn.sofifa.com/flags/es.png   81   
...                                         ...  ...   
259088      https://cdn.sofifa.com/flags/fr.png   69   
259090  https://cdn.sofifa.com/flags/gb-wls.png   66   
259091  https://cdn.sofifa.com/flags/gb-eng.png   67   
259101      https://cdn.sofifa.com/flags/ba.png   73   
259105      https://cdn.sofifa.com/flags/es.png   74   

                        team_&_contract  height  weight   foot  growth  \
id                                                                       
2                   Udinese 2008 ~ 2016     183    81.0   Left       0   
16                KAS Eupen 2014 ~ 2019     178    64.0  Right       0   
27            Coventry City 2016 ~ 2020     175    72.0  Right       0   
36                      Sunderland 2009     180    74.0  


## Some columns we can drop in this step,as we won't need them in analysis or prediction


'Player_photo'- not useful information

'Club_logo' -not useful information

'Flag_photo'- not useful information

'Team_contract'- we have separate 'club' and 'contract' columns,so we can drop this one

'joined'-we have already splitted contract column(contract_start,contract_end).So 'joined' is the same of 

'contract_start'

'Contract'- We split it and drop original

'Gender' -there are only male category

'Loan_date_end'- 16215 from 17125 columns are null

'Position' -We have 'bp'-best position column,so we dont need now this one

In [18]:
df.drop(['player_photo', 'club_logo','flag_photo','team_&_contract','gender','joined','loan_date_end','position',
        'contract'],axis = 1, inplace=True)

In [19]:
# Converting some columns to number,and then rounding all numerical columns

df['height'] = round(df['height'].astype(int))
df['release_clause'] = round(df['release_clause'].astype(int))
df['wage'] = round(df['wage'].astype(int))
df['value'] = round(df['value'].astype(int))

X_num = df.select_dtypes('float64','int64')
for column in X_num.columns:
    df[column] = round(df[column])

In [20]:
#Putting contract_start and contract_end columns forward 

columns = list(df.columns)

columns.insert(16, columns.pop())
columns.insert(16, columns.pop())

df = df[columns]

##  Lets look at our cleaned data

In [21]:
df

name  age  ova            nationality           club  bov  \
id                                                                             
2           G. Pasquale   33   69                  Italy        Udinese   71   
16          Luis García   37   71                  Spain      KAS Eupen   70   
27              J. Cole   33   71                England  Coventry City   71   
36             D. Yorke   36   68  Trinidad &amp; Tobago     Sunderland   70   
41              Iniesta   36   81                  Spain    Vissel Kobe   82   
...                 ...  ...  ...                    ...            ...  ...   
259088       A. Medioub   22   64                 France     CD Tondela   66   
259090       W. Rickard   18   56                  Wales        Burnley   58   
259091       C. Barrett   18   56                England        Burnley   56   
259101   J. Gazibegović   20   62     Bosnia Herzegovina  SK Sturm Graz   62   
259105  Pablo Rodríguez   18   59                  Spain          Lecce   61   

         bp  pot  height  weight   foot  growth    value   wage  \
id                                                                
2       LWB   69     183    81.0   Left       0   625000   7000   
16       CM   71     178    64.0  Right       0   600000   7000   
27      CAM   71     175    72.0  Right       0  1100000  15000   
36       ST   82     180    74.0  Right      14        0      0   
41      CAM   81     170    68.0  Right       0  5500000  12000   
...     ...  ...     ...     ...    ...     ...      ...    ...   
259088   CB   69     198    74.0  Right       5   550000   2000   
259090  CAM   66     178    76.0  Right      10   130000   3000   
259091   RB   67     175    69.0  Right      11   130000   2000   
259101   RB   73     175    67.0  Right      11   475000   3000   
259105   ST   74     178    69.0  Right      15   325000   2000   

        release_clause  attacking contract_start   contract_end  crossing  \
id                                                                          
2                    0        313          2008            2016        75   
16             1100000        337          2014            2019        68   
27                   0        337          2016            2020        80   
36                   0        264           2009           None        54   
41             7200000        367          2018            2021        75   
...                ...        ...            ...            ...       ...   
259088               0        197         Jun 30   2021 On Loan        35   
259090          273000        248          2020            2021        50   
259091          273000        218          2020            2022        55   
259101          784000        223          2020            2023        62   
259105          788000        254          2020            2024        32   

        finishing  heading_accuracy  short_passing  volleys  skill  dribbling  \
id                                                                              
2              50                59             71     58.0    338         73   
16             64                61             76     68.0    369         69   
27             64                41             77     75.0    387         79   
36             70                60             80     45.0    255         68   
41             69                54             90     79.0    408         85   
...           ...               ...            ...      ...    ...        ...   
259088         23                66             47     26.0    193         44   
259090         43                50             64     41.0    264         57   
259091         34                53             42     34.0    212         52   
259101         26                52             60     23.0    258         55   
259105         64                59             45     54.0    235         57   

        curve  fk_accuracy  long_passing  ball_con

#  Looks better,but contract_start and contract_end columns,also the last columns with '+' sign need to be cleaned

In [22]:
# Now we transform cleaned data to sql 

connection = 'mysql+pymysql://root:root@mysql_db:3306/fifa21'
engine = create_engine(connection)
df.to_sql('players', engine,if_exists = 'replace')

ModuleNotFoundError: No module named 'pymysql'